### PR TITLE
Approve token spend page

### LINKED ISSUE #727: Token approvals should have editable amounts on the signature page

### PR DESCRIPTION
**Pull request by wojciech-turek**
When users are approving token spend of an ERC-20 they should see a specific page and be able to change the approval amount.

https://user-images.githubusercontent.com/61434499/148082553-f3c706a8-282c-4fb7-bce2-97cb1c2ae1dc.mov

See #726 
Closes #727

### CODE REVIEW

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx [hunk 1]
```diff
@@ -1,8 +1,67 @@
...
+  const dispatch = useDispatch()
+  const [changing, setChanging] = useState(false)
+  const accountData = useBackgroundSelector(selectCurrentAccountBalances)
+
+  const { assetAmounts } = accountData ?? {
+    assetAmounts: [],
+  }
+
+  // ERC-20 the approval amount will be in the range of 74-138 in form of a 32 bytes hex string
+  const approveAmount = transactionDetails?.input?.substring(74, 138)
```
Comment (mhluongo) [inline ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:29]: I think we should properly parse this with the ABI, ala `EnrichmentService`... but not blocking.

File: ui/pages/SignTransaction.tsx [hunk 2]
```diff
@@ -47,22 +47,26 @@ export default function SignTransaction(): ReactElement {
   const history = useHistory()
   const dispatch = useBackgroundDispatch()
   const location = useLocation<SignLocationState | undefined>()
+  const transactionDetails = useBackgroundSelector(selectTransactionData)
   const { assetSymbol, amount, to, value, signType } = location.state ?? {
-    signType: SignType.Sign,
+    // If tx input begins with bytecode that corresponds to ERC-20 approve
+    // we load the approve page
+    signType: transactionDetails?.input?.startsWith("0x095ea7b")
```
Comment (mhluongo) [inline ui/pages/SignTransaction.tsx:54]: Ditto

**Review by mhluongo — CHANGES_REQUESTED** [review]
The code looks mergeable, but the approval selector isn't coming up for dashboard.keep.network :grimacing:

**Comment by mhluongo** [comment]
I couldn't test on Uniswap (this wallet already had an approval) so I tried the Keep dashboard. Results below

Screenshot from 2021-12-28 10-58-20

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx [hunk 3]
```diff
@@ -1,8 +1,57 @@
...
+  const infiniteApproval =
+    approvalLimit ===
+    "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
+
+  const asset = assetAmounts.find((el) =>
+    "contractAddress" in el.asset
+      ? el.asset.contractAddress === transactionDetails.to
+      : undefined
+  )
+  const hexedAmount = ethers.utils.hexlify(2)
```
[the remark this answers is not in the document]
Comment (wojciech-turek) [reply ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx]: I will create that in this PR

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx (hunk 1 above)
Comment (wojciech-turek) [reply ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:29]: You are right, I am not 100% how to wire it up at this point though.

File: background/lib/utils/index.ts [hunk 4]
```diff
@@ -106,3 +106,28 @@ export function getEthereumNetwork(): EVMNetwork {
...
+  return Number(BigInt(hexValue) / 10n ** BigInt(tokenDecimals)).toString()
+}
+
+export const numberTo32BytesHex = (value: string, decimals: number): string => {
+  const withDecimals = BigInt(value) * 10n ** BigInt(decimals)
+  const hex = utils.hexlify(withDecimals)
+  return utils.hexZeroPad(hex, 32).split("0x")[1]
+}
+
+export const verifyIsInfiniteAmount = (amount: string): boolean => {
```
Comment (mhluongo) [inline background/lib/utils/index.ts:128]: This name is both too specific and not enough! Perhaps `isInfiniteERC20Amount` or `isMaxUint256`?

File: background/services/indexing/index.ts [hunk 5]
```diff
@@ -387,6 +387,22 @@ export default class IndexingService extends BaseService<Events> {
     }
   }
 
+  getTokenMetadata = async (
```
Comment (mhluongo) [inline background/services/indexing/index.ts:390]: I'd recommend

```typescript
async getTokenMetadata(contractAddress: HexString, network: EVMNetwork)
```

and error out if the network isn't what you want, or ignore it and leave a TODO. Multi-network is open in another PR, don't want to keep making this mistake.

File: background/services/indexing/index.ts [hunk 6]
```diff
@@ -387,6 +387,22 @@ export default class IndexingService extends BaseService<Events> {
...
+    contractAddress: string
+  ): Promise<SmartContractFungibleAsset | undefined> => {
+    const provider = this.chainService.pollingProviders.ethereum
+    // pull metadata from Alchemy
+    const customAsset =
+      (await getTokenMetadata(provider, contractAddress)) || undefined
```
Comment (mhluongo) [inline background/services/indexing/index.ts:396]: What if we already have the data in the DB? Shouldn't we check their first?

Either paradigm works, but would be good in docs

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx [hunk 7]
```diff
@@ -1,28 +1,120 @@
...
+    setChanging(!changing)
+    if (changing && transactionDetails && asset?.decimals) {
+      const updatedInput = `${transactionDetails.input?.substring(
+        0,
+        74
+      )}${numberTo32BytesHex(
+        approvalLimit,
+        asset?.decimals
+      )}${transactionDetails.input?.substring(
+        138,
```
Comment (mhluongo) [inline ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:63]: Can we pull this out to a util function? Fine to leave it in the same file (for now), but it's tough to understand tx parsing + rendering intermingled.

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx [hunk 8]
```diff
@@ -1,28 +1,120 @@
...
+          <SharedAssetIcon size="large" symbol={asset?.symbol ?? ""} />
         </div>
       </div>
-      <span className="site">Uniswap</span>
-      <span className="spending_label">Spend KEEP tokens</span>
+      <span className="site">Smart Contract Interaction</span>
+      <span className="spending_label">
+        {asset?.symbol ? (
+          `Spend ${
+            asset?.symbol ?? truncateAddress(transactionDetails.to || "")
```
Comment (mhluongo) [inline ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:84]: Putting on my Antonio hat, I think you want `??`

File: ui/pages/SignTransaction.tsx [hunk 9]
```diff
@@ -50,23 +50,29 @@ export default function SignTransaction({
 
   const history = useHistory()
   const dispatch = useBackgroundDispatch()
-
-  const { assetSymbol, amount, to, value, signType } = location?.state ?? {
-    signType: SignType.Sign,
+  const transactionDetails = useBackgroundSelector(selectTransactionData)
+  const approvalCodes = ["0x095ea7b3", "0xcae9ca51"] // approve, approveAndCall
```
Comment (mhluongo) [inline ui/pages/SignTransaction.tsx:54]: This should also be a util IMO. It'll make refactoring later simpler.

**Review by mhluongo — COMMENTED** [review]
Left a few thoughts, haven't tried the latest code locally though

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx (hunk 1 above)
Comment (Shadowfiend) [reply ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:41]: I'm going to say… It *is* blocking. We should not be doing byte twiddling. Our previous foray, which we let through as not blocking, led to a lot of unexpected issues and some half fixes, including one I did that was completely broken. In general: we have the tools to do this right and should wield them.

I expect the preferred _flow_ here is going to be to pipe through the enrichment service; however, we don't need to solve that full flow here to use the correct parsing strategy. The short version here is we can move [`parseERC20Tx`](https://github.com/tallycash/extension/blob/main/background/services/enrichment/index.ts#L80-L88) to `lib/erc20.ts`, then use it here. That parse will give us an object whose `name` property should be `"approve"` (we should check that) and which should also have a property `args: { spender: string, value: BigNumber }`. The `value` can then be compared against [`ethers.constants.MaxUint256`](https://docs.ethers.io/v5/api/utils/constants/#constants-MaxUint256).

To rewrite the transaction, we can use the `ERC20_INTERFACE` exported by `lib/erc20.ts` to do `ERC20_INTERFACE.encodeFunctionData("approve", { spender: ..., value: updatedValue })`.

**Review by Shadowfiend — CHANGES_REQUESTED** [review]
Just one… Blocker :grimacing:

File: background/services/indexing/index.ts (hunk 6 above)
Comment (wojciech-turek) [reply background/services/indexing/index.ts:396]: This whole thing is gone with the newest iteration.

File: ui/pages/SignTransaction.tsx (hunk 9 above)
Comment (wojciech-turek) [reply ui/pages/SignTransaction.tsx:54]: It doesn't belong there anyway
Alt Text

File: ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx (hunk 7 above)
Comment (wojciech-turek) [reply ui/components/SignTransaction/SignTransactionApproveSpendAssetBlock.tsx:63]: This part was removed and replaced with use of ERC20_INTERFACE, but I know it was ugly :(

**Comment by mhluongo** [comment]
Testing this on the [Keep dashboard](https://dashboard.keep.network/0x4495c5bf67ba43aa279f41c90083bbbfcf46e643/coverage-pools/deposit) and I'm still not seeing an approval page.

Screenshot from 2022-01-14 08-58-07

**Comment by mhluongo** [comment]
Testing on Uniswap and Chrome as the Keep dash appears to use `approveAndCall`. Running into unrelated bugs. @wojciech-turek can you rebase or merge `main`?

**Review by Shadowfiend — APPROVED** [review]
I think we're good to go here, but haven't had a chance to test drive. Once that's done, merge is a go, and the remaining comments here can be handled separately.

**Comment by Shadowfiend** [comment]
Changed closed to see for #726 as it also covers rendering infinite approvals correctly in the transaction list.

**Comment by mhluongo** [comment]
Screenshot from 2022-01-19 15-29-30
Screenshot from 2022-01-19 15-30-08
Screenshot from 2022-01-19 15-30-46

**Pull request merged**